### PR TITLE
freedesktop modules: init

### PR DESCRIPTION
**Pull request by michaelpj**
###### Motivation for this change
Follow-up to/supersedes https://github.com/NixOS/nixpkgs/pull/44866.

This adds some modules that handle various parts of the XDG specs. This allows us to move some of the logic out of `system-path.nix`, and have a more sensible way to decide whether to update MIME caches.

I then added an option to `system-path` to allow extra environment setup commands to be defined by other modules, so the MIME module can also have the setup code in it.

Points I'm unsure about
- Should the XDG modules be enabled by default? Lots of things probably assume they are, but I could see the argument for not doing so. Currently all the DEs turn them on, so off by default would mean that a machine without a DE or X would also not have all the MIME stuff linked in, and so on.
- I wasn't brave enough to try and have a module controlling the use of the base directory spec - this is pretty deeply woven into everything, so I'm not sure that gains us much. However, it would give us a central location to define the constants `/share` and `/etc/xdg`, which might have some value.
- Is the `extraSetup` mechanism for `system-path` OK? It's a bit ugly that clients use `$out` to be the output of another derivation.

@jtojnar @vcunat 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/config/xdg/menus.nix
Comment (jtojnar) [inline nixos/modules/config/xdg/menus.nix]: typo dekstop

**Comment by Mic92** [comment]
it could be also enabled if a display manager is enabled.

File: nixos/modules/programs/environment.nix
Comment (jtojnar) [inline nixos/modules/programs/environment.nix]: Where did this go?

File: nixos/modules/config/system-path.nix
Comment (jtojnar) [inline nixos/modules/config/system-path.nix]: Why this? The rest of the `share` subdirectories in `pathsToLink` would be pointless then. I think we want to avoid `"/share"` (principle of least privilege or something).

Comment (michaelpj) [reply nixos/modules/config/system-path.nix]: Hm, I think I was confused by the need for `/etc/xdg` but not `/share` - I think you're right that it will do the right thing because we typically include subdirectories of `/share` explicitly.

File: nixos/modules/programs/environment.nix
Comment (michaelpj) [reply nixos/modules/programs/environment.nix]: Whoops, left over from when I did try having a base-directories module.

**Comment by michaelpj** [comment]
Hm, good point about non-DE environments. `xserver.nix` also enables them, perhaps that's enough and I should remove the ones in the DE modules? OTOH, not sure how that will play with Wayland.

**Comment by michaelpj** [comment]
Okay, I've removed the settings in the DEs, relying on the xserver module to set the option.

File: nixos/modules/config/xdg/menus.nix
Comment (jtojnar) [inline nixos/modules/config/xdg/menus.nix]: It looks like it should be `share/menus/applications-merged`: https://specifications.freedesktop.org/menu-spec/menu-spec-latest.html#paths

File: nixos/modules/config/xdg/mime.nix
Comment (michaelpj) [reply nixos/modules/config/xdg/mime.nix]: Yes, but I believe that `update-desktop-database` only performs actions relevant to MIME lookup, so I think it belongs here.

File: nixos/modules/config/xdg/menus.nix
Comment (michaelpj) [reply nixos/modules/config/xdg/menus.nix]: Ah yes, good spot.

Comment (michaelpj) [reply nixos/modules/config/xdg/menus.nix]: Fixed.

File: nixos/modules/config/xdg/autostart.nix
Comment (jtojnar) [inline nixos/modules/config/xdg/autostart.nix]: Unused `pkgs` arg

File: nixos/modules/config/xdg/autostart.nix
Comment (jtojnar) [inline nixos/modules/config/xdg/autostart.nix]: It would be nice to have a link to the specification `<link xlink:href="https://specifications.freedesktop.org/autostart-spec/autostart-spec-latest.html">XDG Autostart specification</link>`

File: nixos/modules/config/system-path.nix
Comment (jtojnar) [inline nixos/modules/config/system-path.nix]: Maybe mention something about the `$out` thing.

File: nixos/modules/config/system-path.nix
Comment (jtojnar) [inline nixos/modules/config/system-path.nix]: This should probably be moved to `xdg.icons`

Comment (michaelpj) [reply nixos/modules/config/system-path.nix]: Hm, I wasn't sure about that - I thought maybe there should be a GTK module that does this?

Comment (jtojnar) [reply nixos/modules/config/system-path.nix]: Right, that looks like GTK thing: https://developer.gnome.org/gtk3/stable/gtk-update-icon-cache.html

We should probably get a module later.

Comment (michaelpj) [reply nixos/modules/config/system-path.nix]: Similarly for the glib schemas.

**Comment by michaelpj** [comment]
Comments addressed. I put back the test that the target directory be writable in the fragments. I also fixed a mistake with the `menus` and `menus/applications-merged` directories - those should have been under `/etc/xdg`.

File: nixos/modules/config/xdg/menus.nix
Comment (jtojnar) [inline nixos/modules/config/xdg/menus.nix]: Still some unused `pkgs` args.

Comment (michaelpj) [reply nixos/modules/config/xdg/menus.nix]: Removed.

**Comment by michaelpj** [comment]
Hold off this for a bit while I do some more testing.

**Comment by michaelpj** [comment]
Okay, it turns out that linking `/share` is very important, and many applications rely on it. That seems potentially kind of problematic, but I've left it in for now. Many DEs already provided this setting, and it's likely that anything that didn't was quite broken.

At this point, a diff of my system-path built with and without this reveals only the previously missing XDG menu dirs.

**Comment by michaelpj** [comment]
(so I'm fairly confident in this now)

**Comment by jtojnar** [comment]
> Okay, it turns out that linking `/share` is very important, and many applications rely on it. That seems potentially kind of problematic, but I've left it in for now. Many DEs already provided this setting, and it's likely that anything that didn't was quite broken.

Those applications are broken then. I intend to remove `/share` from GNOME and replace it with specific directories like `/share/gnome-shell/extensions` and `/share/gnome-shell/search-providers`.

**Comment by jtojnar** [comment]
Also, similar to `icons`, we might want to have `sounds`, as per [sound-theme-spec](https://freedesktop.org/wiki/Specifications/sound-theme-spec/) (for `/share/sounds`). And appstream [AppStream](https://www.freedesktop.org/software/appstream/docs/) for `/share/appdata` and `/share/metainfo`.

File: nixos/modules/config/xdg/mime.nix
Comment (michaelpj) [reply nixos/modules/config/xdg/mime.nix]: Aha, good catch. I had thought I could get away with removing them since the use of the binary is now encoded here, but it seems like it also has data to install.

**Comment by michaelpj** [comment]
> Those applications are broken then. I intend to remove /share from GNOME and replace it with specific directories like /share/gnome-shell/extensions and /share/gnome-shell/search-providers.

I think that's a good idea, but I think we can do it as future work.

> Also, similar to icons, we might want to have sounds, as per sound-theme-spec (for /share/sounds). And appstream AppStream for /share/appdata and /share/metainfo.

Sure, but let's do follow-ups. As an aside, it looks like we install a bunch of stuff in `/share/sounds` that doesn't follow the spec.

**Comment by michaelpj** [comment]
I don't think we should leave the linking of `/share` individual DEs. There are a *lot* of applications that have data that doesn't get linked without it, and I suspect things will just be subtly broken if you don't have it.

I don't think we need a doc change, this should only make things work better. However, I forgot to ask: should we consider having the default for some of these modules be *not* enabled? That would mean that systems without X could get away with e.g. not having icons, which might shrink closures a bit. OTOH, then we would *definitely* need a release note, since some things might rely on it happening implicitly.

**Comment by michaelpj** [comment]
For example, I noticed this because `/share/sddm` wasn't linked, causing it to not work properly. That means that if we take it out, anyone using sddm without a DE will be broken. Of course, the correct fix is for the sddm module to link it, but I don't want to do all those fixes here.

**Comment by michaelpj** [comment]
This is the diff between the profiles - as you can see, a lot of stuff goes missing.
```
diff --brief -r /nix/store/9mj6n7ggca7r6d82sdjvjpd4rl8c01m3-system-path /nix/store/1rfgr9cwxh7n535fqv8ldcmahpbj7v0j-system-path | grep "Only in" > missing.txt
```
[missing.txt](https://github.com/NixOS/nixpkgs/files/2293839/missing.txt)

**Comment by jtojnar** [comment]
> I don't think we should leave the linking of `/share` to individual DEs. There are a *lot* of applications that have data that doesn't get linked without it, and I suspect things will just be subtly broken if you don't have it.

Moving it up in the chain would make the future granularization harder as more bugs would be obscured until then. I think status quo is good enough for now.

> should we consider having the default for some of these modules be *not* enabled? That would mean that systems without X could get away with e.g. not having icons, which might shrink closures a bit. OTOH, then we would *definitely* need a release note, since some things might rely on it happening implicitly.

Hmm, 

```ShellSession
$ nix path-info -f . -S shared-mime-info
/nix/store/g89z1wlss6dnhcvajxmn36nbayql80rh-shared-mime-info-1.10	   48303016
$ nix path-info -f . -S desktop-file-utils
/nix/store/i6iglvwcq294nyam49h0q04aqzjzqf2c-desktop-file-utils-0.23	   39642584
```

But looks like it is mostly just `glibc`, `libxml2` and `glib`. Perhaps @vcunat can address this.

> For example, I noticed this because `/share/sddm` wasn't linked, causing it to not work properly. That means that if we take it out, anyone using sddm without a DE will be broken. Of course, the correct fix is for the sddm module to link it, but I don't want to do all those fixes here.

I would leave this as is and address this in further PR.

> This is the diff between the profiles - as you can see, a lot of stuff goes missing.
> ```
> diff --brief -r /nix/store/9mj6n7ggca7r6d82sdjvjpd4rl8c01m3-system-path /nix/store/1rfgr9cwxh7n535fqv8ldcmahpbj7v0j-system-path | grep "Only in" > missing.txt
> ```
> [missing.txt](https://github.com/NixOS/nixpkgs/files/2293839/missing.txt)

In my experience, most apps compile in the path. Only minority looks up their own data in `$XDG_DATA_DIRS`. I can recall quodlibet looking for translations in there because the Python package installation method they use does not support absolute installation paths.

**Comment by michaelpj** [comment]
Hm, I was thinking that we would save on closures by not linking `/share/icons`, but while this would shrink the size of the system environment tree it wouldn't save on the closure unless the icons were in a separate output.

**Comment by michaelpj** [comment]
In case it wasn't clear from my comment before: since there's unlikely to be a closure size improvement, I think we should leave these options enabled by default.

**Review by jtojnar — APPROVED** [review]
Great job. Looks good to me, though I would still like another set of eyes to check this.

**Comment by michaelpj** [comment]
@vcunat could you have a look at the final state of this?

File: nixos/modules/config/system-path.nix
Comment (joachifm) [inline nixos/modules/config/system-path.nix]: Nit: I think using an empty string as the default value is strictly more correct.

**Comment by michaelpj** [comment]
ping

**Comment by jtojnar** [comment]
It is still in some reviewers’ queues. Let’s leave it one more week.

**Comment by michaelpj** [comment]
Wasn't intending to be pushy, just thought we'd got pretty close :)

**Comment by samueldr** [comment]
Don't hang the reviews and merging on this issue, but I'm wondering why the `enable` options have `default = true`, but are then enabled in `xserver.nix`, what would break defaulting them to false? (I'm not sure any of the previous conversations are relevant here.)

**Comment by michaelpj** [comment]
Enabling them in `xserver.nix` is mainly just to signal intent, and so we *could* turn them off if we wanted. As for what would break if we set them to false... I'm not sure, and I'm not brave enough to try it. Especially since the benefits seems slim. Maybe a follow-up PR can try that.

**Review by samueldr — APPROVED** [review]

**Pull request merged**
